allennlp/data/transforms.py

```diff
@@ -0,0 +1,558 @@
+        The key of the ``MetadataField`` indicating what "type" of instance this is.
+    in_metadata : ``boool``: optional (default = False)
+        Whether the partition key actually exists in a Field with the name "metadata",
+        rather than being a field itself.
+    """
+
+    def __init__(self, batch_size: int, partition_key: str = "dataset", in_metadata: bool = False):
+
+        self.batch_size = batch_size
+        self.partition_key = partition_key
+        self.in_metadata = in_metadata
+
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[Batched]:
+
+        instances = list(dataset)
+
+        hoppers: Dict[str, List[Instance]] = defaultdict(list)
+
+        for instance in instances:
+            if self.in_metadata:
+                partition = instance.fields["metadata"].metadata[self.partition_key]  # type: ignore
+            else:
+                partition = instance.fields[self.partition_key].metadata  # type: ignore
+            hoppers[partition].append(instance)
+
+        # Get a `lazy_groups_of` iterator over each set of homogeneous instances.
+        batches = {
+            key: lazy_groups_of(iter(hopper), self.batch_size) for key, hopper in hoppers.items()
+        }
+
+        remaining = set(batches)
+        # Yield batches in a round-robin fashion until none are left.
+        while remaining:
+            for key, lazy_batches in batches.items():
+                if key in remaining:
+                    try:
+                        batch = next(lazy_batches)
+                        yield batch
+                    except StopIteration:
+                        remaining.remove(key)
+
+
+@Transform.register("biggest_batch_first")
+class BiggestBatchFirst(InstanceTransform[Batched]):
+    def transform(self, dataset: Iterable[Instance]) -> Iterable[Batched]:
+
+        raise NotImplementedError(
+            "BiggestBatchFirst must be called on batches. "
+            "Add a transform which batches your instances to your pipeline."
+        )
+
+    def transform_batch(self, batches: Iterable[Batched]) -> Iterable[Batched]:
+
+        batches = list(batches)
+        if len(batches) <= 2:
+            yield from batches
+        else:
+            last = batches.pop()
+            penultimate = batches.pop()
+            yield from [last, penultimate] + batches
+
+
+@Transform.register("shuffle")
+class Shuffle(InstanceTransform[Batched]):
+    """
+    Shuffles a set of instances and returns them.
+
+    NOTE: This could be generic over batches and instances, i.e
+    for the batch case it just shuffles the batches. This is purposely NOT
+    like that, because it makes it easy to accidentally shoot yourself
+    in the foot if you have a dataset which is infinite, because in the batched
+    case, we still try to read all batches into memory. This way,
+    if this is called on a dataset which is already batched, it shuffles within
```

Comment:
Shuffling within a batch accomplishes nothing.  If I sort by padding, how am I supposed to ensure that the batches are shuffled?